org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/launch/TestLaunchShortcut.java

```diff
@@ -0,0 +1,225 @@
+import org.eclipse.buildship.ui.UiPlugin;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Platform;
+import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.debug.ui.ILaunchShortcut;
+import org.eclipse.jdt.core.IClassFile;
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.IMethod;
+import org.eclipse.jdt.core.IType;
+import org.eclipse.jdt.core.ITypeRoot;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.internal.ui.actions.SelectionConverter;
+import org.eclipse.jdt.ui.JavaUI;
+import org.eclipse.jface.text.ITextSelection;
+import org.eclipse.jface.viewers.ISelection;
+import org.eclipse.jface.viewers.ISelectionProvider;
+import org.eclipse.jface.viewers.IStructuredSelection;
+import org.eclipse.ui.IEditorPart;
+
+import com.google.common.base.Function;
+import com.google.common.base.Optional;
+import com.google.common.base.Predicate;
+import com.google.common.collect.FluentIterable;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableList.Builder;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.gradleware.tooling.toolingclient.GradleDistribution;
+
+/**
+ * This {@link ILaunchShortcut} is used to offer the opportunity to run tests
+ * from the editor or package explorer.
+ */
+@SuppressWarnings("restriction")
+public class TestLaunchShortcut implements ILaunchShortcut {
+
+    @Override
+    public void launch(ISelection selection, String mode) {
+        if (selection instanceof IStructuredSelection) {
+            launch(((IStructuredSelection) selection).toList(), mode);
+        } else {
+            logNoTestsFound();
+        }
+    }
+
+    @Override
+    public void launch(IEditorPart editor, String mode) {
+        ITypeRoot element = JavaUI.getEditorInputTypeRoot(editor.getEditorInput());
+        if (element != null) {
+            Optional<IMethod> selectedMethod = resolveSelectedMethodName(editor, element);
+            if (selectedMethod.isPresent()) {
+                launch(ImmutableList.of(selectedMethod.get()), mode);
+            } else {
+                launch(ImmutableList.of(element), mode);
+            }
+        } else {
+            logNoTestsFound();
+        }
+    }
+
+    protected void performLaunch(List<? extends IJavaElement> elementsToLaunch, String mode) {
+        // get the right workingDirExpression for the
+        // GradleRunConfigurationAttributes
+        IJavaElement javaElement = elementsToLaunch.get(0);
+        IProject project = javaElement.getResource().getProject();
+        String workingDirExpression = project.getLocation().toOSString();
+
+        GradleRunConfigurationAttributes configurationAttributes = GradleRunConfigurationAttributes.with(
```

Comment:
@donat Are the GradleRunConfigurationAttributes configured right? I am not sure, how to easily  get the GradleDistribution or other Gradle specific information from an IProject